packages/js/core/src/types/WrapError.ts

```diff
@@ -0,0 +1,334 @@
+import { CleanResolutionStep } from "../algorithms";
+
+/**
+Wrap error codes provide additional context to WrapErrors.
+
+Error code naming convention (approximate):
+    type of handler
+    type of functionality
+    piece of functionality
+    ==> handler_typeFn_pieceFn
+
+Error code map:
+  0-24 -> Client
+  25-49 -> URI resolution
+  50-74 -> Wasm wrapper invocation & sub-invocation
+  75-99 -> Plugin invocation & sub-invocation
+  100-255 -> Unallocated
+ */
+export enum WrapErrorCode {
+  UNKNOWN,
+  CLIENT_LOAD_WRAPPER,
+  CLIENT_GET_FILE,
+  CLIENT_GET_IMPLEMENTATIONS,
+  CLIENT_VALIDATE_RESOLUTION,
+  CLIENT_VALIDATE_ABI,
+  CLIENT_VALIDATE_RECURSIVE,
+  URI_RESOLUTION = 25,
+  URI_RESOLVER,
+  URI_NOT_FOUND,
+  WASM_INVOKE_ABORTED = 50,
+  WASM_SUBINVOKE_ABORTED,
+  WASM_INVOKE_FAIL,
+  WASM_MODULE_NOT_FOUND,
+  WASM_STATE_ERROR,
+  PLUGIN_INVOKE_FAIL = 75,
+  PLUGIN_METHOD_NOT_FOUND,
+  PLUGIN_ARGS_MALFORMED,
+}
+
+export type WrapErrorSource = Readonly<{
```

Comment:
Is this a wasm implementation detail? If so, it should be in the wasm-js library

Reply:
I moved the type declaration to wasm-js and am instead using a type literal in the WrapError class.

It's specific to Wasm wrappers, but it's also necessarily a property of WrapError since WrapError needs to know how to handle both Wasm and plugin errors. Maybe it would make more sense to have something like a core WrapError and then either subclass it or use a composition/dependency injection sort of pattern where the core WrapError holds a more specific type of error that it delegates some behavior to?

Reply:
I think sub-classing a core `WrapError` makes sense to me. This is how it works with custom TypeScript / JavaScript errors.

Reply:
Okay, I'm looking at this now. 

One thing worth considering is that the parsing logic needs to be able to parse any `WrapError` since subinvocations can return any type of `WrapError` string. I think that means the `WrapError` will need to know about all its children, or `wasm-js` will need to know about an error subclass that's defined somewhere like `plugin-js`. Is that okay?

Reply:
I instead added error source info for exceptions thrown in plugin wrappers, making the property more universal.

Example:
```bash
    WrapError: I'm throwing!
    code: 50 WRAPPER INVOKE ABORTED
    uri: wrap://plugin/mock
    method: methodThatThrows
    args: {} 
    source: { file: "/Users/kris/WebstormProjects/monorepo/packages/js/client/src/__tests__/helpers/mockPluginRegistration.ts", row: 13, col: 17 }
```